### PR TITLE
HDDS-7321. Auto rocksDB small sst files compaction.

### PR DESCRIPTION
**Pull request by ChenSammi**
https://issues.apache.org/jira/browse/HDDS-7321

### CODE REVIEW

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/statemachine/DatanodeConfiguration.java [hunk 1]
```diff
@@ -358,6 +358,31 @@ public void setWaitOnAllFollowers(boolean val) {
   )
   private int rocksdbMaxOpenFiles = ROCKSDB_MAX_OPEN_FILES_DEFAULT;
 
+  @Config(key = "rocksdb.auto-compaction-small-sst-file",
```
Comment (myskov) [inline hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/statemachine/DatanodeConfiguration.java:361]: looks like these props should be described in ozone-default.xml

**Comment by myskov** [comment]
BTW, the [HDDS-7332](https://issues.apache.org/jira/browse/HDDS-7332) doesn't say anything about rocksdb compaction

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/statemachine/DatanodeConfiguration.java (hunk 1 above)
Comment (adoroszlai) [reply hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/statemachine/DatanodeConfiguration.java:361]: @myskov these config annotations are generated at compile time into XML format (search for `ozone-default-generated.xml` after build).  They should not be added manually to `ozone-default.xml`.

**Comment by ChenSammi** [comment]
> BTW, the [HDDS-7332](https://issues.apache.org/jira/browse/HDDS-7332) doesn't say anything about rocksdb compaction

Right,  It's HDDS-7331 instead.

**Review by DaveTeng0 — COMMENTED** [review]
LGTM!(one unit test failed! TestOzoneContainer #testBuildNodeReport)

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/statemachine/DatanodeConfiguration.java (hunk 1 above)
Comment (DaveTeng0) [reply hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/statemachine/DatanodeConfiguration.java:361]: hmm just for myself learning~ when do we usually add configuration parameters into file _ozone-default.xml_, and when to add to the _Datanodeconfiguration_ class~?

Comment (ChenSammi) [reply hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/statemachine/DatanodeConfiguration.java:361]: Per my understanding,  annotation style configuration is preferred over ozone-default.xml currently.  @DaveTeng0 .

Comment (DaveTeng0) [reply hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/statemachine/DatanodeConfiguration.java:361]: got it!! Thanks sammi!!

**Comment by symious** [comment]
@ChenSammi Thank you for the work. I think it's a very good feature to have.

Two things come to my mind:
1. The interface we are using is `check()`, but `compaction` seems more than a check operation since it's updating the rocksdb.
2. The performance issue might not be annoying, Users or Devs might not be easy to find their cluster slow caused by this operation. 

I was wondering if we don't do the compaction operation in `check()`, but only add a `need to compact` tag. Meanwhile, we can have a compaction thread doing the real compaction when Datanode is not busy so that the Datanode won't be affected by the compaction.

**Comment by ChenSammi** [comment]
@symious , thanks for the review.  

>     1. The interface we are using is `check()`, but `compaction` seems more than a check operation since it's updating the rocksdb.
> 
>     2. The performance issue might not be annoying, Users or Devs might not be easy to find their cluster slow caused by this operation.
> 
> 
> I was wondering if we don't do the compaction operation in `check()`, but only add a `need to compact` tag. Meanwhile, we can have a compaction thread doing the real compaction when Datanode is not busy so that the Datanode won't be affected by the compaction.

Rocksdb has implemented somehow the way you suggested.  rocksdb.compactRange will return immediately. The real compaction will happen in another background thread per RocksDB's schedule.  

About the performance concern, I have did some calculation about the rocksdb size of single volume. Say volume is 16TB, in small object/file situation, say 1MB each file, there will be 16 million files(one block each file).  Currently, one block with one chunk is 63 byte(protobuf serialized), it's totally about 1GB for all block data.  Consider there is other data in rocksdb, the total size of rocksdb per 16TB volume could be no more than several GB, or less than 1GB if the files are big files.  So I'm not very worry about the rocksdb performance on datanode.

**Comment by DaveTeng0** [comment]
hello guys~ looks like sammi has addressed all questions. Should we merge this if no further comment?

**Comment by symious** [comment]
I found there is a new value for this option: "ManagedCompactRangeOptions.BottommostLevelCompaction" which is "kForceOptimized", the comment for this value is as follows.
```
  // Always compact bottommost level but in bottommost level avoid
  // double-compacting files created in the same compaction
```

I'm not sure if this new value is brought up because of some performance issue, may be we can also make this configurable for the convenience of future upgrade?

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/statemachine/DatanodeConfiguration.java [hunk 2]
```diff
@@ -364,6 +364,31 @@ public void setWaitOnAllFollowers(boolean val) {
...
+      defaultValue = "true",
+      type = ConfigType.BOOLEAN,
+      tags = { DATANODE },
+      description = "Auto compaction small sst files when condition is met. "
```
Comment (kerneltime) [inline hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/statemachine/DatanodeConfiguration.java:371]:
```suggestion
      description = "Auto compact small SST files (rocksdb.auto-compaction-small-sst-file-size-threshold) when count exceeds (rocksdb.auto-compaction-small-sst-file-num-threshold)"
```

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/statemachine/DatanodeConfiguration.java [hunk 3]
```diff
@@ -364,6 +364,31 @@ public void setWaitOnAllFollowers(boolean val) {
...
+  )
+  private boolean autoCompactionSmallSstFile = true;
+
+  @Config(key = "rocksdb.auto-compaction-small-sst-file-size-threshold",
+      defaultValue = "1MB",
+      type = ConfigType.SIZE,
+      tags = { DATANODE },
+      description = "Sst files with size smaller than threshold is a target."
```
Comment (kerneltime) [inline hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/statemachine/DatanodeConfiguration.java:379]:
```suggestion
      description = "SST files smaller than this configuration will be auto compacted."
```

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/statemachine/DatanodeConfiguration.java [hunk 4]
```diff
@@ -364,6 +364,31 @@ public void setWaitOnAllFollowers(boolean val) {
...
+  )
+  private long autoCompactionSmallSstFileSize = 1024 * 1024;
+
+  @Config(key = "rocksdb.auto-compaction-small-sst-file-num-threshold",
+      defaultValue = "512",
+      type = ConfigType.INT,
+      tags = { DATANODE },
+      description = "Auto compaction will happen if the number of small Sst " +
+          " files exceeds this threshold."
```
Comment (kerneltime) [inline hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/statemachine/DatanodeConfiguration.java:388]:
```suggestion
      description = "Auto compaction will happen if the number of small SST " +
          " files exceeds this threshold."
```

**Review by kerneltime — COMMENTED** [review]
It looks good, I need to check the usage of the underlying Rocks DB interfaces.
cc @duongkame @prashantpogde

**Comment by symious** [comment]
> Cannot find this kForceOptimized in the latest RocksDB release 7.9.2. Can you share the link?

The option is in https://github.com/facebook/rocksdb/blob/main/include/rocksdb/options.h#L1844, seems not added to jni yet, since it's added in 2019, not sure why it's not included in jni.

I raised a PR there, https://github.com/facebook/rocksdb/pull/11181, hoping someone can make some comments.

Anyway, I think it will be good to have the option configurable, maybe in the future, there could be some new choice on the options.

**Comment by ChenSammi** [comment]
> > Cannot find this kForceOptimized in the latest RocksDB release 7.9.2. Can you share the link?
> 
> The option is in https://github.com/facebook/rocksdb/blob/main/include/rocksdb/options.h#L1844, seems not added to jni yet, since it's added in 2019, not sure why it's not included in jni.
> 
> I raised a PR there, [facebook/rocksdb#11181](https://github.com/facebook/rocksdb/pull/11181), hoping someone can make some comments.
> 
> Anyway, I think it will be good to have the option configurable, maybe in the future, there could be some new choice on the options.

@symious , thanks for the further info. Configuration properties are for customers to do the tuning by themselves.  As most customers don't need to know or have the knowledge of these low level specific rocksdb properties meaning and impact, let's choose a good default value instead as current.

**Comment by symious** [comment]
> let's choose a good default value instead as current.

Ok, it makes sense.

**Comment by ChenSammi** [comment]
Thanks @symious @kerneltime for the code review.  All the comments are addressed. Please help to take another look of it?

**Comment by symious** [comment]
@ChenSammi Thank you for the contribution.
@myskov @kerneltime @DaveTeng0 Thank you for the review.

**Comment by ChenSammi** [comment]
Thanks @symious @myskov @kerneltime @DaveTeng0 for the code review.

**Pull request merged**